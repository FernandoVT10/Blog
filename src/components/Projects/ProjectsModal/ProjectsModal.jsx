import Modal from "../../Modal";
import ProjectCard from "../ProjectCard/";
import Api from "../../../ApiController";
import Loader from "../../Loader/";

import { useState, useEffect, useRef } from "react";
import { useRouter } from "next/router";

export default () => {
    const [modalActive, setModalActive] = useState(false);
    const [projects, setProjects] = useState([]);
    const [loading, setLoading] = useState(false);
    const [skill, setSkill] = useState(null);
    const modalProjectIsActive = useRef();

    const router = useRouter();

    useEffect(() => {
        const querySkill = router.query.skill;
        const queryProject = router.query.project;

        if(queryProject) {
            modalProjectIsActive.current = true;
        } else {
            modalProjectIsActive.current = false;
        }
        
        if(querySkill && !queryProject) {
            if(skill === querySkill) {
                setModalActive(true);
            } else {
                setSkill(querySkill);
                setLoading(true);

                Api.get(`projects/getProjectsBySkillName/${querySkill}`)
                .then(projects => {
                    setProjects(projects);
                    setModalActive(true);
                    setLoading(false);
                });
            }
        } else {
            setModalActive(false);
        }
    }, [router.query]);

    const onCloseModal = () => {
        if(!modalProjectIsActive.current) {
            router.push({
                pathname: router.pathname,
                query: {}
            });
        } else {
            // when the modal is closed, we add the modal-open class to the body
            document.body.classList.add("modal-open");
        }
    };

    const getProjects = () => {
        if(loading) {
            return (
                <Loader active={true} />
            );
        }

        if(!projects.length) {
            return (
                <div className="col-12">
                    <h4 className="color-secondary text-center font-weight-bold">
                        There is not available projects
                    </h4>
                </div>
            );
        }

        return projects.map(project => {
            return (
                <div className="col-12 col-lg-6 mb-3" key={project._id}>
                    <ProjectCard project={project} />
                </div>
            );
        });
    };

    return (
        <Modal
        title={`${router.query.skill} Projects`}
        active={modalActive}
        modalSize="modal-xl"
        onClose={onCloseModal}
        prefix="projects">
            <div className="container-fluid p-0">
                <div className="row">
                    { getProjects() }
                </div>
            </div>
        </Modal>
    );
}