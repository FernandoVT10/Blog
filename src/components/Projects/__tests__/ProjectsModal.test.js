import ProjectsModal from "../ProjectsModal/";
import { act, Simulate } from "react-dom/test-utils";
import { render } from "react-dom";

const PROJECTS_MOCK = [
    {
        _id: 123,
        title: "Test title",
        images: ["test-image-1.jpg", "test-image-2.jpg", "test-image-3.jpg"]
    },
    {
        _id: 124,
        title: "Test title 2",
        images: ["test-image-1.jpg", "test-image-2.jpg", "test-image-3.jpg"]
    },
    {
        _id: 125,
        title: "Test title 3",
        images: ["test-image-1.jpg", "test-image-2.jpg", "test-image-3.jpg"]
    }
];

window.$ = () => ({ modal: jest.fn(), on: jest.fn() });

jest.mock("next/router", () => ({
    useRouter: jest.fn(() => ({ query: { skill: "React JS" } })),
}));

let container;

test("Check if <ProjectsModal/> renders correctly", async () => {
    fetchMock.doMock();

    fetchMock.mockOnce(JSON.stringify(
        PROJECTS_MOCK
    ));

    container = document.createElement('div');
    document.body.appendChild(container);

    await act(async () => {
        render(<ProjectsModal/>, container);
    });

    expect(container).toMatchSnapshot();
});