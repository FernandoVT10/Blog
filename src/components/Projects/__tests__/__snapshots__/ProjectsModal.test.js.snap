// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if <ProjectsModal/> renders correctly 1`] = `
<div>
  <div
    class="modal fade"
    id="projects-modal"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl"
      role="document"
    >
      <div
        class="modal-content modal__content"
      >
        <div
          class="modal-header modal__header"
        >
          <h5
            class="modal__title"
          >
            React JS Projects
          </h5>
          <button
            aria-label="Close"
            class="close modal__button"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          class="modal-body modal__body"
        >
          <div
            class="container-fluid p-0"
          >
            <div
              class="row"
            >
              <div
                class="col-12 col-lg-6 mb-3"
              >
                <div
                  class="project"
                >
                  <img
                    alt="Test title"
                    class="project__image project__image--active"
                    src="/img/projects/test-image-1.jpg"
                  />
                  <img
                    alt="Test title"
                    class="project__image "
                    src="/img/projects/test-image-2.jpg"
                  />
                  <img
                    alt="Test title"
                    class="project__image "
                    src="/img/projects/test-image-3.jpg"
                  />
                </div>
              </div>
              <div
                class="col-12 col-lg-6 mb-3"
              >
                <div
                  class="project"
                >
                  <img
                    alt="Test title 2"
                    class="project__image project__image--active"
                    src="/img/projects/test-image-1.jpg"
                  />
                  <img
                    alt="Test title 2"
                    class="project__image "
                    src="/img/projects/test-image-2.jpg"
                  />
                  <img
                    alt="Test title 2"
                    class="project__image "
                    src="/img/projects/test-image-3.jpg"
                  />
                </div>
              </div>
              <div
                class="col-12 col-lg-6 mb-3"
              >
                <div
                  class="project"
                >
                  <img
                    alt="Test title 3"
                    class="project__image project__image--active"
                    src="/img/projects/test-image-1.jpg"
                  />
                  <img
                    alt="Test title 3"
                    class="project__image "
                    src="/img/projects/test-image-2.jpg"
                  />
                  <img
                    alt="Test title 3"
                    class="project__image "
                    src="/img/projects/test-image-3.jpg"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
